### ISSUE TITLE
Tables: CombiTimeTable with spline interpolation and linear extrapolation

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1028]
**Reported by beutlich on 14 Mar 2013 12:34 UTC**
Consider the following table model
```
model Unnamed
  Modelica.Blocks.Sources.CombiTimeTable t_old(
    table=[0.3,0;0.4,1;0.5,0;0.6,-1;0.7,0],
    smoothness=Modelica.Blocks.Types.Smoothness.ContinuousDerivative);
end Unnamed;
```
Should the extrapolated values of intervals [0,0.3) and (0.7,1] be taken from the boundary spline polynoms or from the linear extrapolation?

![Image](Unnamed.png)

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1028

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 26 Mar 2013 10:33 UTC**
According to the documentation of CombiTimeTable (and also to common sense, since extrapolation with polynomials is usually not what the user desires), the correct behavior is linear extrapolation outside of the definition domain (the Dymola behavior is an error).

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 26 Mar 2013 19:17 UTC**
This ticket is invalid as it's really an issue of Dymola and not the MSL. And hence there was nothing to fix here.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 26 Mar 2013 19:18 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 26 Mar 2013 20:38 UTC**
The expected behavior (see `t_new.y[1]` above) is already implemented in the revised implementation of the MSL tables (and I guess will make it in MSL 3.2.1).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 19 Jun 2013 10:25 UTC**
Looking more I believe that neither the previous behavior nor the current one is the correct one (yes, should have found this earlier).

The smoothInterpolation guarantees that you have continuous first order derivatives **inside** the interval; and it seems natural to have at least a **choice** for extrapolation that guarantees this for the **entire** interval. 

If you want a first order extrapolation outside the interval the correct approach would naturally be to keep the derivative-value at the end-point.

```
Modelica.Blocks.Sources.CombiTimeTable combiTimeTable1(table=[0.0,8; 1,-1; 2,
        6], smoothness=Modelica.Blocks.Types.Smoothness.ContinuousDerivative);
```
simulate 0 to 3 seconds and note that we get a discontinuous first order derivative at time=2.

```imagebox
[[Image(DerTest.png, 600px)]]
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 19 Jun 2013 10:48 UTC**
At least the current behaviour is correct according to the documentation
```
Extrapolate linearly through the last two table points outside of the table scope
```
and the name _Extrapolation.LastTwoPoints_. ;-)
So if I will try to fix this, should we add a new option _LastPointAndSlope_ to _Extrapolation_ enum or handle this implicitly?

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 19 Jun 2013 11:13 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 20 Jun 2013 15:04 UTC**
attachment:T1028.patch solves this extrapolation issue on source level for all four external tables. 1D case is pretty clean, but 2D case really is a mess. I am open for any better implementation suggestions. That's the reason I did not yet commit the changes. Thank you.

TODO:
 * Comment correction of _Extrapolation.LastTwoPoints_
 * Documentation adaptions

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 21 Jun 2013 13:49 UTC**
Resolved by aa9a4ae54580586b4147653fedf02ff6de98488d (1D case), 7b1c1a5f5ae0507be71d6c2acd6cfb691377a9c4 (2D case) and ddf5648f171310577284f24010de96943dbbbff7 (comment, doc).

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 26 Jun 2013 09:18 UTC**
